lib/install/management/upgrade.go

```diff
@@ -77,97 +79,112 @@ func (d *Dispatcher) Upgrade(vch *vm.VirtualMachine, conf *config.VirtualContain
+		d.retryDeleteSnapshot(snapshotName, conf.Name)
 		return nil
 	}
 	log.Errorf("Failed to upgrade: %s", err)
 	log.Infof("Rolling back upgrade")
 
-	// reset timeout, to make sure rollback still happens in case of deadline exceeded error in previous step
-	var cancel context.CancelFunc
-	d.ctx, cancel = context.WithTimeout(context.Background(), settings.RollbackTimeout)
-	defer cancel()
-
-	if rerr := d.rollback(conf, snapshotName); rerr != nil {
+	if rerr := d.rollback(conf, snapshotName, settings); rerr != nil {
 		log.Errorf("Failed to revert appliance to snapshot: %s", rerr)
-		// return the error message for upgrade, instead of rollback
 		return err
 	}
+	log.Infof("Appliance is rolled back to old version")
 
 	d.deleteUpgradeImages(ds, settings)
-	log.Infof("Appliance is rollback to old version")
+	d.retryDeleteSnapshot(snapshotName, conf.Name)
+	// return the error message for upgrade
 	return err
 }
 
-func (d *Dispatcher) deleteSnapshot(id types.ManagedObjectReference, snapshotName string, applianceName string) error {
-	defer trace.End(trace.Begin(snapshotName))
-	log.Infof("Deleting upgrade snapshot %q", snapshotName)
-	// do clean up aggressively, even the previous operation failed with context deadline excceeded.
-	d.ctx = context.Background()
-	if _, err := d.appliance.WaitForResult(d.ctx, func(ctx context.Context) (tasks.Task, error) {
-		return d.appliance.RemoveSnapshot(ctx, id, true, true)
-	}); err != nil {
+// retryDeleteSnapshot will retry to delete snpashot if there is GenericVmConfigFault returned. This is a workaround for vSAN delete snapshot
+func (d *Dispatcher) retryDeleteSnapshot(snapshotName string, applianceName string) error {
+	// delete snapshot immediately after snapshot rollback usually fail in vSAN, so have to retry several times
+	operation := func() error {
+		return d.deleteSnapshot(snapshotName, applianceName)
+	}
+	var err error
+	if err = retry.Do(operation, isSystemError); err != nil {
 		log.Errorf("Failed to clean up appliance upgrade snapshot %q: %s.", snapshotName, err)
 		log.Errorf("Snapshot %q of appliance virtual machine %q MUST be removed manually before upgrade again", snapshotName, applianceName)
-		return err
 	}
-	return nil
+	return err
 }
 
-func (d *Dispatcher) createSnapshot(name string, desc string) (*types.ManagedObjectReference, error) {
-	defer trace.End(trace.Begin(name))
-	log.Infof("Creating snapshot %s", name)
+func isSystemError(err error) bool {
+	if soap.IsSoapFault(err) {
+		switch soap.ToSoapFault(err).VimFault().(type) {
+		case types.SystemError:
+			return true
+		}
+	}
 
-	snapRefID, err := d.tryCreateSnapshot(name, desc)
-	if err == nil {
-		log.Infof("created snapshot %s", snapRefID)
-		return snapRefID, nil
+	if soap.IsVimFault(err) {
+		log.Infof("Is soap vim fault")
+		switch soap.ToVimFault(err).(type) {
+		case *types.SystemError:
+			return true
+		}
 	}
-	log.Error(err)
-	return nil, err
+
+	if terr, ok := err.(task.Error); ok {
```

Comment:
Nit-pick: since we're only looking for one `case` in all type switch assertions in this func, we can replace them with `if` assertions like on line 136.